include/boost/url/impl/const_string.hpp

```diff
@@ -101,149 +47,83 @@ const_string(
     Allocator const& a)
 {
     char* dest;
-    p_ = construct(
-        s.size(), a, dest);
-    std::memcpy(dest,
-        s.data(), s.size());
+    *this = const_string(s.size(), a, dest);
+    std::memcpy(dest, s.data(), s.size());
 }
 
 const_string::
 const_string(
     const_string const& other) noexcept
     : string_view(other)
-    , p_(other.p_)
+    , refs_(other.refs_)
+    , alloc_(other.alloc_)
 {
-    if(p_)
-        ++p_->refs;
+    if (!increment_refs())
+    {
+        *this = const_string(
```

Comment:
we are calling a function that can throw from a constructor marked noexcept. We shouldn't be trying to "handle" the case where the counter overflows. Just use a 32-bit unsigned counter.

Reply:
> We shouldn't be trying to "handle" the case where the counter overflows

Why not? 

- This saves more than 7 bytes per string in the pool: the size bytes and alignment bytes. Not to mention what we have already saved from `impl`.
- Even most strings are usually going to be shorter than what we just saved here. 
- If we promise the library avoids allocations, pool memory is also much more valuable than heap memory. Users could be disappointed if they need a pool larger than the strings themselves or even a pool that takes more than it needs to.
- There's no practical use case for more than 255 copy-like references of the string. If the user gets to this point, they are either doing something wrong or don't mind an extra string every 255 copies.
- uint8_t allows the reference counter and the string to be aligned, which simplifies the allocation logic immensely: a single allocator, no rebinding, no over_allocator, not anything.
- It can't be smaller than uint8_t, which makes it less arbitrary and much easier to justify.
- `uint8_t` is the only size for which there is a qualitative difference in the pros and cons. For any other size without handling overflow (16, 32, 64), we lose all the benefits, the cons all apply, and any difference is only quantitative, making any limit (2^16, 2^32, 2^64) as justifiable as the other.

Reply:
You need to remove the noexcept from the constructor if you do this, because throwing when noexcept calls `std::terminate`.

Reply:
Oh... sure. No problem.

Reply:
I'm not sure this is correct. Suppose you get to the part in `increment_refs` where the old value is 255, and then the thread is suspended. The counter is now 0. Suppose now that three threads make copies. The counter is now 3. The thread is resumed, does fetch_sub, gets the value to 2. Doesn't seem good.

Using compare_exchange(_strong) would make it correct but probably less efficient. I'm not sure whether it's possible to make it correct using only ++/--.

Reply:
You're right. The old rationale was that only one thread was going to return 0 anyway so this could be relaxed. I didn't consider the effects of the final counter being wrong.   

So I realized the best pattern here is probably the same we use for decrementing the counter, with `fetch_sub` instead of `deallocate`. Not much better than acq_rel in practice, but it seems like the correct implementation.

Reply:
use a 32-bit counter and dont get fancy with copying

Reply:
This is not a memory order problem; it happens with seq_cst as well.

Reply:
we are prematurely optimizing again, 255 is too small for the maximum refs, and it causes us to have to include ridiculous code to copy the string which is not justifiable without overwhelming field experience, of which we have none.